File: compute/ei_winfree/neuron.py
import random
import numpy as np
from matplotlib import pyplot as plt
from pprint import pprint
import sys
import os
from datetime import datetime

import params as prm


N = prm.N
gamma = prm.gamma
r = prm.r
use_cpp = prm.use_cpp
tmpfile_key = f'{datetime.now()}'.replace(' ', '-').replace(':', '-').replace('.', '-')
tmpi = f'tmp/i{tmpfile_key}.tmp'
tmpo = f'tmp/o{tmpfile_key}.tmp'

#%% prepare omega
omega_bar_E = prm.omega_bar_E
omega_bar_I = prm.omega_bar_I
K_EE = prm.K_EE
K_II = prm.K_II
K_EI = prm.K_EI
K_IE = prm.K_IE


def omega_from_lorentzian(gamma, omega_bar):
    y = random.uniform(0, 1)
    omega = gamma * np.tan(np.pi*(y-1/2)) + omega_bar
    return omega

omega_E = []
omega_I = []
for i in range(N):
    omega_E.append(omega_from_lorentzian(gamma, omega_bar_E))
    omega_I.append(omega_from_lorentzian(gamma, omega_bar_I))
    
omega_E = np.array(omega_E)
omega_I = np.array(omega_I)

#%% prepare theta
theta_E = np.array([random.uniform(0, np.pi*2) for i in range(N)])
theta_I = np.array([random.uniform(0, np.pi*2) for i in range(N)])

#%% theta_dot
def P(theta):
    return (1-r)*(1+np.cos(theta))/(1-2*r*np.cos(theta)+r**2)

def Q(theta):
    return 1 - np.cos(theta)

def h(theta_E1, theta_I1):
    return 1/N * np.sum(P(theta_E1)), 1/N * np.sum(P(theta_I1))

def theta_dot(theta_E1, theta_I1):
    h_E, h_I = h(theta_E1, theta_I1)
    return omega_E + Q(theta_E1) * (K_EE * h_E - K_EI * h_I), omega_I + Q(theta_I1) * (K_IE * h_E - K_II * h_I)

#%% simulation
dt = prm.dt
t_max = prm.t_max

t = 0
i = 0
step = 17

tr = []
hEr = []
hIr = []

if use_cpp:
    print(f'using cpp, temp file key = {tmpfile_key}')
    dat = '' #N, K_EE, K_II, K_EI, K_IE, r, theta_E, theta_I, omega_E, omega_I, t_max, dt
    dat = f'{N}\n{K_EE}\n{K_II}\n{K_EI}\n{K_IE}\n{r}\n{t_max}\n{dt}\n\n\n'
    for theta in theta_E:
        dat += f'{theta}\n'
    dat += '\n'
    for theta in theta_I:
        dat += f'{theta}\n'
    dat += '\n'
    for omega in omega_E:
        dat += f'{omega}\n'
    dat += '\n'
    for omega in omega_I:
        dat += f'{omega}\n'
    dat += '\n'
    with open(tmpi, 'w+') as f:
        f.write(dat)

    os.system(f'./neuron {tmpfile_key}')
    
    with open(tmpo, 'r') as f:
        dat_s = f.read()
    dat1 = dat_s.split('/')
    datt = dat1[0].split()
    dathE = dat1[1].split()
    dathI = dat1[2].split()
    for k in range(len(datt)):
        tr.append(float(datt[k]))
        hEr.append(float(dathE[k]))
        hIr.append(float(dathI[k]))
else:
    while t < t_max:
        kE1, kI1 = theta_dot(theta_E, theta_I)
        kE2, kI2 = theta_dot(theta_E + dt/2 * kE1, theta_I + dt/2 * kI1)
        kE3, kI3 = theta_dot(theta_E + dt/2 * kE2, theta_I + dt/2 * kI2)
        kE4, kI4 = theta_dot(theta_E + dt * kE3, theta_I + dt * kI3)
        
        theta_E = theta_E + dt/6 * (kE1 + 2*kE2 + 2*kE3 + kE4)
        theta_I = theta_I + dt/6 * (kI1 + 2*kI2 + 2*kI3 + kI4)
        t += dt
        i += 1

        if i % step == 0:
            sys.stdout.write(f'\r {t}     ')
            tr.append(t)
            h_E, h_I = h(theta_E, theta_I)
            hEr.append(h_E)
            hIr.append(h_I)

print()
tr = np.array(tr)
hEr = np.array(hEr)
hIr = np.array(hIr)

plt.scatter(tr, hEr, s=1)
plt.scatter(tr, hIr, s=1)
plt.xlabel('t')
plt.ylabel('$h_E, h_I$')
plt.show()

File: compute/ei_winfree/neuron.cpp
#include <iostream>
#include <fstream>
#include <cmath>

#define MAX 100000

using namespace std;

int N;
double K_EE, K_II, K_EI, K_IE;
double r;
double t_max, dt;

double theta_E[MAX], theta_I[MAX];
double omega_E[MAX], omega_I[MAX];

double P(double theta);
double Q(double theta);
void h(double* theta_E, double* theta_I, double* h);
void theta_dot(double* theta_E, double* theta_I, double* kE, double* kI);


int main(int argc, char** argv) {
	
	string tmpfile_key(argv[1]);
	string tmpi = "tmp/i" + tmpfile_key + ".tmp";
	string tmpo = "tmp/o" + tmpfile_key + ".tmp";

	ifstream icache(tmpi);
	icache >> N;
	icache >> K_EE >> K_II >> K_EI >> K_IE;
	icache >> r;
	icache >> t_max >> dt;
	
	//cout << "K_EE = " << K_EE << endl;
	//cout << "r = " << r << endl;
	//cout << "K_EI = " << K_EI << endl;
	
	for(int i=0; i<N; ++i) icache >> theta_E[i];
	for(int i=0; i<N; ++i) icache >> theta_I[i];

	for(int i=0; i<N; ++i) icache >> omega_E[i];
	for(int i=0; i<N; ++i) icache >> omega_I[i];
	
	icache.close();

	double kE1[N], kI1[N], kE2[N], kI2[N], kE3[N], kI3[N], kE4[N], kI4[N];
	double thE1[N], thI1[N], thE2[N], thI2[N], thE3[N], thI3[N];
	
	double t = 0;
	int i = 0, step = 17;
	int est_recs = int(t_max / dt / step) + 10, real_recs = 0;
	
	double hEr[est_recs], hIr[est_recs], tr[est_recs];
	double hEI[2];
	

	while(t < t_max) {
		
		theta_dot(theta_E, theta_I, kE1, kI1);
		for(int j=0; j<N; ++j) {thE1[j] = theta_E[j] + dt/2 * kE1[j]; thI1[j] = theta_I[j] + dt/2 * kI1[j];}
		theta_dot(thE1, thI1, kE2, kI2);
		for(int j=0; j<N; ++j) {thE2[j] = theta_E[j] + dt/2 * kE2[j]; thI2[j] = theta_I[j] + dt/2 * kI2[j];}
		theta_dot(thE2, thI2, kE3, kI3);
		for(int j=0; j<N; ++j) {thE3[j] = theta_E[j] + dt * kE3[j]; thI3[j] = theta_I[j] + dt * kI3[j];}
		theta_dot(thE3, thI3, kE4, kI4);

		for(int j=0; j<N; ++j) {
			theta_E[j] += dt/6 * (kE1[j] + 2*kE2[j] + 2*kE3[j] + kE4[j]);
			theta_I[j] += dt/6 * (kI1[j] + 2*kI2[j] + 2*kI3[j] + kI4[j]);
		}

		t += dt;
		i += 1;
		
		if (i % step == 0) {
			h(theta_E, theta_I, hEI);
			hEr[real_recs] = hEI[0];
			hIr[real_recs] = hEI[1];
			tr[real_recs] = t;
			real_recs += 1;
			cout << "\r" << t << flush;
		}
	}


	ofstream ocache(tmpo);
	for(int k=0; k<real_recs; ++k) {
		ocache << tr[k] << endl;
	}
	ocache << '/' << endl;
	for(int k=0; k<real_recs; ++k) {
		ocache << hEr[k] << endl;
	}
	ocache << '/' << endl;
	for(int k=0; k<real_recs; ++k) {
		ocache << hIr[k] << endl;
	}

	ocache.close();


}

double Q(double theta) {
	return 1 - cos(theta);
}

double P(double theta) {
	return (1-r)*(1+cos(theta)) / (1-2*r*cos(theta)+r*r);
}

void h(double* theta_E1, double* theta_I1, double* hEI) {
	
	double sumE = 0, sumI = 0;

	for(int i=0; i<N; ++i) {
		sumE += P(theta_E1[i]);
		sumI += P(theta_I1[i]);
	}
		
	hEI[0] = sumE / N;
	hEI[1] = sumI / N;


}

void theta_dot(double* theta_E1, double* theta_I1, double* kE, double* kI) {
	
	double hEI[2];
	h(theta_E1, theta_I1, hEI);
	double hE = hEI[0], hI = hEI[1];

	for(int i=0; i<N; ++i) {
		kE[i] = omega_E[i] + Q(theta_E1[i]) * (K_EE * hE - K_EI * hI);
		kI[i] = omega_I[i] + Q(theta_I1[i]) * (K_IE * hE - K_II * hI);
	}

}
